### PR TITLE
Refactor EeaSendRawTransaction

### PR DESCRIPTION
**Pull request by pinges**
Signed-off-by: Stefan Pingel <stefan.pingel@consensys.net>

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcErrorConverter.java [hunk 1]
```diff
@@ -44,7 +44,7 @@ public static JsonRpcError convertTransactionInvalidReason(
       case GAS_PRICE_TOO_LOW:
         return JsonRpcError.GAS_PRICE_TOO_LOW;
       case PRIVACY_GROUP_DOES_NOT_EXIST:
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcErrorConverter.java]: NIT: If we are renaming the JsonRpcError, should we also rename this enum value to be consistent?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java [hunk 2]
```diff
@@ -115,10 +115,15 @@
   VALUE_NOT_ZERO(-50100, "We cannot transfer ether in private transaction yet."),
   DECODE_ERROR(-50100, "Unable to decode the private signed raw transaction"),
   GET_PRIVATE_TRANSACTION_NONCE_ERROR(-50100, "Unable to determine nonce for account in group."),
-  PRIVACY_GROUP_DOES_NOT_EXIST(-50100, "Privacy group does not exist."),
+  OFFCHAIN_PRIVACY_GROUP_DOES_NOT_EXIST(-50100, "OffChain Privacy group does not exist."),
+  ONCCHAIN_PRIVACY_GROUP_DOES_NOT_EXIST(-50100, "OnChain Privacy group does not exist."),
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java]: I believe `Offchain` and `Onchain` are the spelling we use across the board.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java [hunk 3]
```diff
@@ -284,16 +286,13 @@ public PrivacyGroup retrievePrivacyGroup(final String privacyGroupId, final Stri
       final RLPInput rlpInput =
           RLP.input(privateTransactionSimulatorResultOptional.get().getOutput());
       if (rlpInput.nextSize() > 0) {
-        return decodeList(rlpInput.raw());
+        return new PrivacyGroup(
+            privacyGroupId.toBase64String(), Type.ONCHAIN, "", "", decodeList(rlpInput.raw()));
       } else {
-        return Collections.emptyList();
+        return null;
```
Comment (lucassaldanha) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java]: Can we change this method to return an Optional?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java [hunk 4]
```diff
@@ -245,7 +246,8 @@ public long determineBesuNonce(
   }
 
   @Override
-  public PrivacyGroup retrievePrivacyGroup(final String privacyGroupId, final String enclaveKey) {
+  public PrivacyGroup retrieveOffChainPrivacyGroup(
```
Comment (lucassaldanha) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java]: Can we change this method to return an Optional?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java [hunk 5]
```diff
@@ -77,87 +72,100 @@ public String getName() {
...
 
       final String enclavePublicKey =
           enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser());
-      final ValidationResult<TransactionInvalidReason> validationResult =
-          privacyController.validatePrivateTransaction(privateTransaction, enclavePublicKey);
-      if (!validationResult.isValid()) {
-        return new JsonRpcErrorResponse(
-            id, convertTransactionInvalidReason(validationResult.getInvalidReason()));
+
+      if (!privateTransaction.getPrivateFrom().equals(Bytes.fromBase64String(enclavePublicKey))) {
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java]: Can `privateTransaction.getPrivateFrom()` be null? Do we need to be careful here?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java [hunk 6]
```diff
@@ -77,87 +72,100 @@ public String getName() {
...
-                  enclavePublicKeyProvider.getEnclaveKey(requestContext.getUser()));
-        } catch (final EnclaveClientException e) {
-          // it is an onchain group
+      PrivacyGroup privacyGroup = null;
+      final Optional<Bytes> maybePrivacyGroupId = privateTransaction.getPrivacyGroupId();
+      if (onchainPrivacyGroupsEnabled) {
+        if (!maybePrivacyGroupId.isPresent()) {
+          // on-chain private transaction does not contain privacy group ID
+          return new JsonRpcErrorResponse(
+              id, JsonRpcError.PRIVACY_GROUP_ID_NOT_AVAILABLE_WHILE_ON_CHAIN_PRIVACY_IS_ENABLED);
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java]: We probably don't need this comment. The JsonRpcError name could be a bit simpler I think:
```
JsonRpcError.PRIVACY_GROUP_ID_NOT_AVAILABLE
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java [hunk 7]
```diff
@@ -77,87 +72,100 @@ public String getName() {
...
-        } else {
-          if (isOnchainPrivacyGroupEnabled) {
-            return new JsonRpcErrorResponse(id, OFFCHAIN_PRIVACY_GROUP_NOT_ENABLED);
-          }
-
-          privacyMarkerTransaction =
-              privacyController.createPrivacyMarkerTransaction(enclaveKey, privateTransaction);
+            && !privacyController.isGroupAdditionTransaction(privateTransaction)) {
+          // on-chain privacy enabled, but on-chain privacy group does not exist and is not an
+          // on-chain privacy group creation
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java]: I think we can remove this comment. The code is clear on what is happening.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java [hunk 8]
```diff
@@ -60,15 +66,46 @@ public String getName() {
...
+
+      if (!privateTransaction.getPrivateFrom().equals(Bytes.fromBase64String(enclavePublicKey))) {
+        return new JsonRpcErrorResponse(id, ENCLAVE_ERROR);
+      }
+
+      PrivacyGroup privacyGroup = null;
+      final Optional<Bytes> maybePrivacyGroupId = privateTransaction.getPrivacyGroupId();
+      if (onchainPrivacyGroupsEnabled) {
+        if (!maybePrivacyGroupId.isPresent()) {
+          // on-chain private transaction does not contain privacy group ID
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java]: No need for this comment :)

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java [hunk 9]
```diff
@@ -60,15 +66,46 @@ public String getName() {
...
+          return new JsonRpcErrorResponse(
+              id, JsonRpcError.PRIVACY_GROUP_ID_NOT_AVAILABLE_WHILE_ON_CHAIN_PRIVACY_IS_ENABLED);
+        }
+        privacyGroup =
+            privacyController.retrieveOnChainPrivacyGroup(
+                maybePrivacyGroupId.get(), enclavePublicKey);
+        if (privacyGroup == null
+            && !privacyController.isGroupAdditionTransaction(privateTransaction)) {
+          // on-chain privacy enabled, but on-chain privacy group does not exist and is not an
+          // on-chain privacy group creation
```
Comment (lucassaldanha) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java]: No need for this comment :)

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcErrorConverter.java (hunk 1 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcErrorConverter.java]: Done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java (hunk 2 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java]: Done

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java (hunk 3 above)
Comment (pinges) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java]: done

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java (hunk 4 above)
Comment (pinges) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/DefaultPrivacyController.java]: done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java (hunk 5 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java]: I think the PrivateTransaction.readFrom does not return null ..

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java (hunk 6 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java:95]: Done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java (hunk 7 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java:103]: Done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java (hunk 8 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java:87]: Done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java (hunk 9 above)
Comment (pinges) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivDistributeRawTransaction.java:97]: Done

**Pull request merged**